README.md
# Maru

> REST-like API micro-framework for elixir inspired by [grape](https://github.com/ruby-grape/grape).

[![Build Status](https://img.shields.io/travis/elixir-maru/maru.svg?style=flat-square)](https://travis-ci.org/elixir-maru/maru)
[![Hex.pm Version](https://img.shields.io/hexpm/v/maru.svg?style=flat-square)](https://hex.pm/packages/maru)
[![Docs](https://inch-ci.org/github/elixir-maru/maru.svg?branch=master&style=flat-square)](https://inch-ci.org/github/elixir-maru/maru)
[![Hex.pm Downloads](https://img.shields.io/hexpm/dt/maru.svg?style=flat-square)](https://hex.pm/packages/maru)

## Installation

To get started with Maru, add the following to `mix.exs`:

```elixir
def deps() do
  [
    {:maru, "~> 0.14"},
    {:plug_cowboy, "~> 2.0"},

    # Optional dependency, you can also add your own json_library dependency
    # and config with `config :maru, json_library, YOUR_JSON_LIBRARY`.
    {:jason, "~> 1.1"}
  ]
end
```

## Usage

lib/my_app/server.ex:

```elixir
defmodule MyApp.Server do
  use Maru.Server, otp_app: :my_app
end

defmodule Router.User do
  use MyApp.Server

  namespace :user do
    route_param :id do
      get do
        json(conn, %{user: params[:id]})
      end

      desc "description"

      params do
        requires :age, type: Integer, values: 18..65
        requires :gender, type: Atom, values: [:male, :female], default: :female

        group :name, type: Map do
          requires :first_name
          requires :last_name
        end

        optional :intro, type: String, regexp: ~r/^[a-z]+$/
        optional :avatar, type: File
        optional :avatar_url, type: String
        exactly_one_of [:avatar, :avatar_url]
      end

      # post do
      #   ...
      # end
    end
  end
end

defmodule Router.Homepage do
  use MyApp.Server

  resources do
    get do
      json(conn, %{hello: :world})
    end

    mount Router.User
  end
end

defmodule MyApp.API do
  use MyApp.Server

  before do
    plug Plug.Logger
    plug Plug.Static, at: "/static", from: "/my/static/path/"
  end

  plug Plug.Parsers,
    pass: ["*/*"],
    json_decoder: Jason,
    parsers: [:urlencoded, :json, :multipart]

  mount Router.Homepage

  rescue_from Unauthorized, as: e do
    IO.inspect(e)

    conn
    |> put_status(401)
    |> text("Unauthorized")
  end

  rescue_from [MatchError, RuntimeError], with: :custom_error

  rescue_from :all, as: e do
    conn
    |> put_status(Plug.Exception.status(e))
    |> text("Server Error")
  end

  defp custom_error(conn, exception) do
    conn
    |> put_status(500)
    |> text(exception.message)
  end
end
```

In your `Application` module, add `Server` as a worker:

```elixir
defmodule MyApp.Application do
  use Application

  def start(_type, _args) do
    children = [
      MyApp.Server
    ]

    opts = [strategy: :one_for_one, name: MyApp.Supervisor]
    Supervisor.start_link(children, opts)
  end
end
```

Then configure `maru`:

```elixir
# config/config.exs
config :my_app, MyApp.Server,
  adapter: Plug.Cowboy,
  plug: MyApp.API,
  scheme: :http,
  port: 8880

config :my_app,
  maru_servers: [MyApp.Server]
```

For more information, check out  [Guides](https://maru.readme.io) and [Examples](https://github.com/elixir-maru/maru_examples)


CHANGELOG.md
## Changelog

[Upgrade Instructions](https://maru.readme.io/v0.10/docs/upgrade-instructions-from-v09-1) From v0.9 to v0.10

[Upgrade Instructions](https://maru.readme.io/v0.11/docs/upgrade-instructions-from-v010) From v0.10 to v0.11

[Upgrade Instructions](https://maru.readme.io/v0.12/docs/upgrade-instructions-from-v011) From v0.11 to v0.12

## v0.14.0-dev
* Enhancements
  * Support plug v1.7
* Bugfix
  * fix dialyzer errors
* Deprecations
  * deprecate start maru server by `Maru.Supervisor`

## v0.13.2 (2018-08-23)
* Enhancements
  * Check `json_library` compile time
  * add `Maru.Server` to run server with a standalone mode
* Bugfix
  * `mix maru.routes` error
* Deprecations
  * deprecate `Confex` from maru codebase, in favor of using it in `init/2` of server module
  * soft deprecate start server by `Maru.Supervisor`, in favor of `Maru.Server`

## v0.13.1 (2018-04-20)
* Bugfix
  * take `path_params` back

## v0.13.0 (2018-04-19)
* Enhancements
  * upgrade to plug v1.5
  * make `:json_library ` configurable and use `Jason` as default
* Deprecations
  * remove phoenix parameter support

## v0.12.5 (2017-10-18)
* Enhancements
  * support path_params

## v0.12.4 (2017-10-17)
* Enhancements
  * make regexp validator support list
  * support mount alias modules
  * add `all_or_none_of` validation
* Bugfix
  * Support :chunk response

## v0.12.3 (2017-08-19)
* Bugfix
  * Support confex ~> 3.2

## v0.12.2 (2017-08-08)
* Enhancements
  * log the error instead of raise error when the router module defined but bot loaded
* Bugfix
  * Don't rely on `Mix` module being present

## v0.12.1 (2017-07-26)
* Enhancements
  * add elixir ~> 1.3 and ~> 1.4 support back

## v0.12.0 (2017-07-25)
* Enhancements
  * support set version by `accept` header
  * refactor unittest
  * return original value for `map` and `list` type without `do block`
  * fix elixir 1.5 warning
* Deprecations
  * only support elixir ~> 1.5

## v0.11.4 (2017-03-18)
* Enhancements
  * make all configuration options configurable via OS ENVs, support by [confex](https://hex.pm/packages/confex)
* Bugfix
  * receive messages sent by plug for unittest

## v0.11.3 (2017-01-06)
* Enhancements
  * make `params` a variable instead of macro
  * import module when use a module as helper

## v0.11.2 (2016-11-28)
* Bugfix
  * fix `rescue_from` can't catch `Maru.Exceptions.NotFound` exception
  * fix elixir v1.4 warnings

## v0.11.1 (2016-11-27)
* Bugfix
  * fix typo of exception name: `MethodNotAllow` -> `MethodNotAllowed`, `InvalidFormatter` -> `InvalidFormat`

## v0.11.0 (2016-11-26)
* Enhancements
  * make unittest easier
  * add `with_exception_handlers` option to Maru.Test
  * allow maru to ignore `MIX_ENV=test` or force `test` for other `MIX_ENV` by `config :maru, test: TRUE_OR_FALSE`
  * warning when mount unavailable module
  * warning unknown options for `use Maru.Router` and `use Maru.Test`

## v0.10.6 (2016-11-20)
* Enhancements
  * new `Parameter.Information.type` for one line list parameter `{:list, "MARU.TYPE"}`

## v0.10.5 (2016-10-30)
* Enhancements
  * add `:keep_blank` option for params
  * add `:given` DSL for dependent parameters
  * make `rescue_from` works for all maru router
  * support `with` option for `rescue_from` DSL
  * bring maru's params parser to phoenix

* Bugfix
  * allow set optional params to false and blank value by default

## v0.10.4 (2016-8-19)
* Enhancements
  * support one-line nested list params
  * new DSLs for unittest

## v0.10.3 (2016-7-11)
* Enhancements
  * add detail and responses for description
  * add response helper functions `put_maru_conn/1` and `get_maru_conn/0`
  * pass modified conn to `rescue_from` block

## v0.10.2 (2016-6-28)
* Bugfix
  * fix v1.3.0 exception warning

## v0.10.1 (2016-6-14)
* Bugfix
  * jumbled routes order
  * custom type and validation error

## v0.10.0 (2016-6-11)
* Enhancements
  * totally rewrite route logic
  * totally rewrite params parsing logic and DSLs
  * add overridable plug
  * support top-level plug
  * split Route and Endpoint

* Deprecations
  * `coercion` is deprecated in favor of `type`

## v0.9.6 (2016-4-5)
* Bugfix
  * define helper functions for all environments except `:prod`

## v0.9.5 (2016-3-15)
* Enhancements
  * `mutually_exclusive`, `exactly_one_of` and `at_least_one_of` support `:above_all`

* Bugfix
  * param options for route_param DSL

* Deprecated
  * Maru.Parsers forked from plug

## v0.9.4 (2016-3-5)
* Enhancements
  * Allow param type definition in route_param.

* Bugfix
  * parse bool type error.

## v0.9.3 (2016-1-29)
* Enhancements
  * `maru.routers` is deprecated in favor of `maru.routes`
  * make poison 1.5 and 2.0 compatible

## v0.9.2 (2016-1-6)
* Enhancements
  * no longer keep nil value for optional params

## v0.9.1 (2016-1-4)
* Enhancements
  * update to elixir v1.2

* Bugfix
  * floats can be negative

## v0.9.0 (2015-11-29)
* Enhancements
  * import `Plug.Conn` for Maru.Router by default
  * reuse `text` `html` and `json` to make response

* Bugfix
  * `maru.routers` raise module undefined

* Deprecations
  * return NOT Plug.Conn struct by endpoint
  * custom response helpers: `assigns` `assign` `headers` `header` `content_type` `status` `present` in faver of functions of Plug.Conn.

## v0.8.5 (2015-11-4)
* Bugfix
  * params parser error within `Maru.Test`

## v0.8.4 (2015-10-14)
* Enhancements
  * support rename parameter using `source`
  * fork plug params for reusing request body

## v0.8.3 (2015-10-3)
* Enhancements
  * add paramer type `Json`
  * support parameter coercion using `coerce_with`

## v0.8.2 (2015-9-26)
* Documentation
  * add documentation
  * add ex_doc

* Deprecations
  * remove unused plug: Maru.Plugs.Forword

## v0.8.1 (2015-9-21)
* Enhancements
  * add `&Maru.Builder.Routers.generate/1` to generate endpoints details

## v0.8.0 (2015-9-20)
* Enhancements
  * Update to poison v1.5

## v0.7.2 (2015-9-20)
* Enhancements
  * remove deprecated functions

mix.exs
defmodule Maru.Mixfile do
  use Mix.Project

  def project do
    [
      app: :maru,
      name: "Maru",
      version: "0.14.0-pre.1",
      elixir: "~> 1.4",
      build_embedded: Mix.env() == :prod,
      start_permanent: Mix.env() == :prod,
      deps: deps(),
      description: "REST-like API micro-framework for elixir inspired by grape.",
      source_url: "https://github.com/elixir-maru/maru",
      homepage_url: "https://maru.readme.io",
      package: package(),
      docs: [
        extras: ["README.md"],
        main: "readme"
      ]
    ]
  end

  def application do
    [applications: [:ranch, :cowboy, :plug]]
  end

  defp deps do
    [
      {:plug, "~> 1.7"},
      {:plug_cowboy, "~> 1.0 or ~> 2.0", optional: true},
      {:jason, "~> 1.0 or ~> 1.1", optional: true},
      {:inch_ex, "~> 1.0", only: :docs},
      {:ex_doc, "~> 0.19", only: :docs}
    ]
  end

  defp package do
    %{
      maintainers: ["Xiangrong Hao"],
      licenses: ["BSD 3-Clause"],
      links: %{"Github" => "https://github.com/elixir-maru/maru"}
    }
  end
end
